File: packages/gg-sandbox-python/src/gg/scope.py
"""The names a program is given: the capability modules, and the types they speak in.

Two things are built here, and neither of them varies with the run any more — which is the whole of
gg's capability model as this guest now sees it.

**The modules are the SDK's.** `files`, `board`, `views`, `session` — one per
`gg.catalogue.MODULE_ORDER` entry, carrying every function that module declares, whatever this run
enabled and whatever role this agent holds. A withheld function is an ordinary attribute of an
ordinary namespace, and calling it reaches the host.

**The enforcement is the host's, and it always was.** This SDK is an ordinary Python package: a
program can `import gg.files` and reach every function in it, so a surface built from the run could
never have been a gate — it was only ever a *hint*, and a hint that made a withheld call raise
`AttributeError` from Python rather than a refusal from gg. The membrane refuses a call this agent
was not granted, with a `ToolError` naming the capability that is missing and why, and that refusal
is a value a program can catch and act on. Building the surface from the run bought nothing that the
refusal does not buy better, and cost the model the one thing it needed: a sentence.

**The types are the SDK's too.** Every class a capability module exports is bound, because a type is
not a capability: `TurnRange` is what `context.archive_thread` takes, `ToolError` is what every call
raises and `TaskStatus.DONE` is what a status argument is, and a program that could not name them
would have to reach for a string. `UNCHANGED` travels with them for the same reason.

Everything below is therefore built **once, at import**, into the pre-initialised heap
`componentize-py` snapshots, and a turn pays nothing for it.
"""

from __future__ import annotations

from types import ModuleType
from typing import Any

from . import board, context, core, delegation, files, memories, programs
from . import session as session_module
from . import shell, skills, tasks, views
from ._registry import ATTRIBUTE, REGISTRY
from .catalogue import GG_TOOLS, MODULE_ORDER, PACKAGE, TOOL_BOUND

_MODULES: dict[str, ModuleType] = {
    "files": files,
    "shell": shell,
    "board": board,
    "tasks": tasks,
    "memories": memories,
    "views": views,
    "context": context,
    "delegation": delegation,
    "skills": skills,
    "programs": programs,
    "session": session_module,
    "core": core,
}
"""Every capability module, by its gg module id.

Imported unconditionally, and eagerly. The component is baked once, so there is nothing to gain by
importing lazily — and an import `componentize-py` never executed is a module that is not in the
artifact at all. Importing them is also what populates
`gg._registry.REGISTRY`, since a decorator runs when its module does.
"""


class Bound:
    """A namespace gg assembled rather than one declared in a file.

    The two below share it so that `shim.py` can tell a name that is not part of gg's surface at all
    from an ordinary `AttributeError` against anything else. Python raises the same exception for
    both, and the namespace the failure happened on is the only thing that distinguishes them.
    """

    __slots__ = ()


class Surface(Bound):
    """The `gg` name a program starts with: every capability module, under its id.

    It is what makes a fully-qualified name a thing a program can *write*. `gg.files.read_file` is
    the key a documentation view is opened by, the name search returns, and the name gg quotes back
    at a model in a refusal — so it has to resolve in a program's own scope rather than only after an
    `import`, or the one string the model reads everywhere would be the one string it cannot type.

    The modules on it are the same objects the short names are, so `gg.files.read_file` and
    `files.read_file` are one function. The `gg` **package** is importable too and reaches the same
    functions by a third route; that is deliberate and is not a hole, since nothing here is a gate.
    """

    __slots__ = ("_modules",)

    def __init__(self, modules: dict[str, CapabilityModule]) -> None:
        self._modules = modules

    def __getattr__(self, name: str) -> Any:
        try:
            return self._modules[name]
        except KeyError:
            raise AttributeError(
                f"`{PACKAGE}.{name}` is not one of gg's capability modules; "
                f"they are {', '.join(sorted(self._modules))}",
                name=name,
                obj=self,
            ) from None

    def __dir__(self) -> list[str]:
        return sorted(self._modules)

    def __repr__(self) -> str:
        return f"<gg: {', '.join(sorted(self._modules))}>"


class CapabilityModule(Bound):
    """One of the capability modules a program calls gg through: `files`, `board`, `views`, ….

    A plain namespace would have done, and this exists for what it says when a program reaches for a
    function that is **not** on it — which, now that the surface is the whole SDK, means a name gg
    does not have at all rather than one this run withheld. The difference between
    `'types.SimpleNamespace' object has no attribute 'read_fil'` and a sentence naming the module and
    listing what it declares is a turn.

    It is also what lets the shim classify that mistake as the unknown name it is: Python raises
    `AttributeError` where a guest that could withhold a *name* would raise `NameError`, so the
    failure carries the namespace it happened on and `shim.py` reads it back off the exception.

    It is a **view of** the real module rather than a replacement for it: everything on it is the
    function the module declares, so `files.read_file` and `gg.files.read_file` are one object.
    """

    __slots__ = ("_path", "_members")

    def __init__(self, path: str, members: dict[str, Any]) -> None:
        self._path = path
        self._members = members

    def __getattr__(self, name: str) -> Any:
        try:
            return self._members[name]
        except KeyError:
            raise AttributeError(
                f"`{self._path}.{name}` is not one of the functions gg declares there; "
                f"it declares {', '.join(sorted(self._members))}",
                name=name,
                obj=self,
            ) from None

    def __dir__(self) -> list[str]:
        return sorted(self._members)

    def __repr__(self) -> str:
        return f"<gg `{self._path}`: {', '.join(sorted(self._members))}>"


def _build_modules() -> dict[str, CapabilityModule]:
    """Every capability module, carrying every function this SDK declares on it.

    `gg._registry.REGISTRY` is filled by the `@operation` decorators as the modules above are
    imported, so what a module ends up carrying is exactly what its declarations claim — an operation
    named nowhere is simply absent, which is how `bound_tools` reports a tool the artifact does not
    implement rather than binding a name that would fail when it was called.

    A module with no function on it is not built, which today means `core` alone: it declares the
    types every other module's signatures name and nothing a program calls.
    """
    offered: dict[str, dict[str, Any]] = {}
    for operation, function in REGISTRY.items():
        module = operation.split(".", 1)[0]
        offered.setdefault(module, {})[function.__name__] = function
    return {
        module: CapabilityModule(f"{PACKAGE}.{module}", offered[module])
        for module in MODULE_ORDER
        if module in offered
    }


def _type_names() -> dict[str, Any]:
    """Every type this SDK declares, by the name a program writes.

    The set is read off each module's `__all__`, which is Python's own declaration of a module's
    public surface and the same list the reflector catalogues from. A name that is a function is not
    a type and is skipped: those are the capability modules' business, above.
    """
    bound: dict[str, Any] = {}
    for module in _MODULES.values():
        for name in module.__all__:
            declared = getattr(module, name)
            if getattr(declared, ATTRIBUTE, None) is None:
                bound[name] = declared
    return bound


_MODULE_SURFACE: dict[str, CapabilityModule] = _build_modules()
"""The capability modules, built once at import and therefore inside the baked heap."""

_SCOPE: dict[str, Any] = {
    **_type_names(),
    **_MODULE_SURFACE,
    PACKAGE: Surface(_MODULE_SURFACE),
}
"""Every name a program starts with, likewise built once.

Each module is bound twice, under its bare id and on the `gg` surface, and the two are the same
object: `files.read_file` is the short way and `gg.files.read_file` is the fully-qualified name the
documentation is keyed by, so whichever of them a model reads is one it can write.

`lib` is not here: it is bound by the shim, from the code modules the host handed over, and a run
with none has no such name.
"""


def bound_tools() -> list[str]:
    """The gg tool names this component can bind.

    gg calls this export in a unit test and asserts set-equality with its own `ALL_TOOL_NAMES`. It is
    the one drift gate that inspects the **committed artifact** rather than a source file, so it
    catches the failure no compiler can: a tool added, renamed or removed in gg, with a stale `.wasm`
    still checked in.
    """
    # A gg tool is DISPATCHED by the operation that shares its key — `files.write_file` dispatches
    # `write_file`. The other two rows of `TOOL_BOUND` are a helper and a view that a tool merely
    # *buys*, and reporting either as a tool would put a name in this answer that gg's own vocabulary
    # does not hold.
    dispatched = {
        tool
        for operation, tool in TOOL_BOUND.items()
        if operation in REGISTRY and operation.split(".", 1)[1] == tool
    }
    return [tool for tool in GG_TOOLS if tool in dispatched]


def build_scope() -> dict[str, Any]:
    """Every name a program starts with — a fresh mapping over the same objects.

    A copy rather than the shared dict, because `exec` writes a program's own globals into whatever
    it is handed and one turn's leftovers must not become the next turn's scope. The values are
    shared: they are the SDK's modules and types, and there is one of each.
    """
    return dict(_SCOPE)

File: packages/gg-sandbox-ruby/src/gg/scope.rb
# frozen_string_literal: true

module GG
  # The surface a program is handed: the capability modules, carrying every function this SDK
  # declares.
  #
  # **Nothing here varies with the run, and nothing here is a gate.** Every declaration
  # `GG::Surface` recorded is bound onto its module at load time — inside the heap
  # `componentize-js` snapshots — and stays there for every program of every run. A call this agent
  # was not granted is an ordinary Ruby call that reaches the host, and what comes back is a
  # `GG::Core::ToolError` naming the capability that is missing and why: a value a `rescue` clause
  # can catch and act on, where a `NoMethodError` from Opal was only ever a dead end.
  #
  # The gate is the **host's**, from gg's own operations table, and it is the same one that decides
  # what a `docs.search` will show this agent. There is one reading of "may this agent call X" and
  # it is not in this file.
  #
  # **The types are not built at all**, for the same reason they never were: a type is not a
  # capability. `GG::Core::ToolError` is what a `rescue` clause catches and
  # `GG::Tasks::TaskStatus::DONE` is what a status argument is, and both are declared by this SDK
  # once, into the pre-initialised heap, where they stay.
  #
  # ## Why the implementations are still lifted off and put back
  #
  # Not for gating — for **arity and keywords**. Opal lowers keyword arguments to a trailing hash
  # and silently ignores keys the target does not declare, so `create_issue(reviewer: [...])` — the
  # singular/plural typo — would otherwise be accepted and file an issue with no reviewer, a failure
  # the model cannot see. Each implementation is therefore lifted off its module, its calling shape
  # reflected once, and put back behind a forwarder that checks both. All of it happens at load, so
  # a turn pays nothing.
  #
  # @api private
  module Scope
    # The capability modules, resolved from the names `GG::Surface` orders them by.
    MODULES = Surface::MODULES.map { |name| GG.const_get(name) }.freeze

    # Each declaration's implementation and calling shape, lifted off its module at load time.
    #
    # `componentize-js` runs this file's top level under `wizer` and snapshots the heap, so what is
    # built here is *in the artifact* and costs a turn nothing. Reflecting a method's parameters per
    # turn instead is not free and was measured: over the thirty-five bound functions, twice each,
    # it cost about 3 ms of the roughly 9 ms a whole turn takes — a third of it, on the arm whose
    # entire point is what a language costs.
    IMPLEMENTATIONS = {}

    # The module paths a program may reach, in presentation order — read by `shim.js` for the
    # unknown-name hint, and constant because the surface is.
    MODULE_PATHS = []

    # Lift every module function off the module that declares it, record its implementation and its
    # calling shape, and put it straight back behind the forwarder that checks both. Called at load
    # time; see {IMPLEMENTATIONS}.
    #
    # A member function is left exactly where it is: it is a second way to reach an operation the
    # module function already carries, it hangs off a value a program can only be holding because
    # that operation ran, and lifting it would mean rebinding an unbound instance method for no
    # gain.
    #
    # @return [void]
    def self.lift
      declared = Surface.registry.select(&:module_function?)
      declared.each do |entry|
        callable = entry.owner.method(entry.name)
        IMPLEMENTATIONS[entry] = [callable, keywords(callable), arity(callable)].freeze
        entry.owner.singleton_class.send(:remove_method, entry.name)
      end
      # Only where there is a surface to be outside of. `GG::Core` declares types and no function
      # at all, so a name it does not have is an ordinary Ruby mistake, and pointing at a directory
      # it never carries would be an invitation to a call that does not exist.
      MODULES.select { |mod| Surface.declared_on(mod).any? }.each { |mod| refuse_unknown(mod) }
      IMPLEMENTATIONS.freeze
      declared.each { |entry| bind(entry) }
      MODULES.select { |mod| declared.any? { |entry| entry.owner.equal?(mod) } }
             .each { |mod| MODULE_PATHS << mod.name.to_s }
      MODULE_PATHS.freeze
    end

    # Answer a name a module does not declare with the module, the name, and what it does declare.
    #
    # Every gg function is bound, so a name that lands here is one gg does not have — a typo, or a
    # call out of another arm's vocabulary. The difference between `undefined method 'read_fil' for
    # GG::Files` and a sentence listing what that module declares is a turn.
    #
    # @param mod [Module] the capability module to install the refusal on
    # @return [void]
    def self.refuse_unknown(mod)
      mod.singleton_class.send(:define_method, :method_missing) do |name, *_args|
        declares = Surface.declared_on(mod).map(&:name).sort
        raise NoMethodError,
              "`#{mod.name}.#{name}` is not one of the names gg declares there; " \
              "it declares #{declares.join(', ')}"
      end
      mod.singleton_class.send(:define_method, :respond_to_missing?) do |name, include_private = false|
        Surface.declared_on(mod).any? { |entry| entry.name.to_s == name.to_s } ||
          super(name, include_private)
      end
    end

    # The keyword arguments a lifted implementation declares.
    #
    # @param callable [Method] the implementation
    # @return [Array<String>, nil] the declared keyword names, or nil when it takes `**` and there
    #   is therefore no set to be outside of
    def self.keywords(callable)
      declared = callable.parameters
      return nil if declared.any? { |(kind, _name)| kind == :keyrest }

      declared.select { |(kind, _name)| kind == :key || kind == :keyreq }
              .map { |(_kind, name)| name.to_s }
    end

    # How many positional arguments a lifted implementation takes, as `[minimum, maximum]`.
    #
    # A `nil` maximum is a splat: `GG::Context.archive_thread(*ranges)` takes any number.
    #
    # @param callable [Method] the implementation
    # @return [Array(Integer, Integer, nil)] the fewest and the most it accepts
    def self.arity(callable)
      minimum = 0
      maximum = 0
      callable.parameters.each do |(kind, _name)|
        case kind
        when :req
          minimum += 1
          maximum += 1 unless maximum.nil?
        when :opt
          maximum += 1 unless maximum.nil?
        when :rest
          maximum = nil
        end
      end
      [minimum, maximum]
    end

    # Refuse a positional count the target cannot take, naming the call the way the program wrote
    # it.
    #
    # Opal's own `arity_check` — which gg compiles this SDK with — raises here too, but it says two
    # things this cannot: it renders Ruby's *negative arity encoding* (`expected -3`) for anything
    # with an optional argument, which is a number no CRuby message ever prints. This runs first and
    # says `expected 1..2`.
    #
    # @param fn [String] the call as the program wrote it (`GG::Files.read_file`)
    # @param arity [Array(Integer, Integer, nil)] what {arity} reported for the target
    # @param given [Integer] how many positional arguments arrived
    # @return [void]
    # @raise [ArgumentError] in CRuby's own words, naming the call
    def self.check_arity(fn, arity, given)
      minimum, maximum = arity
      return if given >= minimum && (maximum.nil? || given <= maximum)

      expected = if maximum.nil? then "#{minimum}+"
                 elsif minimum == maximum then minimum.to_s
                 else "#{minimum}..#{maximum}"
                 end
      raise ArgumentError,
            "wrong number of arguments (given #{given}, expected #{expected}) — `#{fn}`"
    end

    # Refuse a keyword the target does not declare, the way CRuby would.
    #
    # Opal lowers keyword arguments to a trailing hash and never reads the extra keys, so
    # `GG::Board.create_issue(…, reviewer: ["r1"])` — the singular/plural typo — is otherwise
    # accepted, and the issue is created with `reviewers: []` while the program believes it named
    # one. {check_arity} does not cover this half: it counts positionals, and an unknown keyword is
    # invisible to it.
    #
    # @param fn [String] the call as the program wrote it (`GG::Board.create_issue`)
    # @param accepted [Array<String>, nil] what the target declares; nil accepts anything
    # @param kwargs [Hash] the keywords the program passed
    # @return [void]
    # @raise [ArgumentError] naming the unknown keywords and the accepted set
    def self.check_keywords(fn, accepted, kwargs)
      return if accepted.nil?

      unknown = kwargs.keys.map(&:to_s) - accepted
      return if unknown.empty?

      # Leading colons written by hand for the reason `GG::Check.choice` writes them by hand: Opal's
      # Symbol IS a String, so `:reviewers.inspect` is `"reviewers"` here, and a message that quoted
      # the accepted set that way would be telling a model to write the one thing this argument does
      # not take.
      takes = if accepted.empty?
                "takes no keyword arguments"
              else
                "takes #{accepted.map { |key| ":#{key}" }.join(", ")}"
              end
      raise ArgumentError,
            "unknown keyword#{unknown.size == 1 ? "" : "s"}: " \
            "#{unknown.map { |key| ":#{key}" }.join(", ")} — `#{fn}` #{takes}"
    end

    # The gg tool names this component can bind.
    #
    # gg calls this export in a unit test and asserts set-equality with its own `ALL_TOOL_NAMES`. It
    # is the one drift gate that inspects the **committed artifact** rather than a source file, so
    # it catches the failure no compiler can: a tool added, renamed or removed in gg, with a stale
    # `.wasm` still checked in.
    #
    # @return [Array<String>] every gg tool this SDK really defines a method for, in declaration
    #   order
    def self.bound_tools
      Surface.registry.select { |entry| IMPLEMENTATIONS.key?(entry) }
             .reject(&:aliased)
             .filter_map(&:tool)
             .uniq
    end

    # Bind one declaration back onto the module that declares it, behind the forwarder that says
    # what Ruby would have said.
    #
    # Forwarded explicitly rather than bound with `define_method(name, &callable)`, which reads
    # better and silently drops the caller's block — so `GG::Files.write_file(path) { … }` would
    # arrive with no body and fail on an argument the program did supply.
    #
    # @param entry [Surface::Entry] the declaration to bind
    # @return [void]
    def self.bind(entry)
      callable, accepted, arity = IMPLEMENTATIONS.fetch(entry)
      fn = "#{entry.owner.name}.#{entry.name}"
      entry.owner.singleton_class.send(:define_method, entry.name) do |*args, **kwargs, &block|
        Scope.check_arity(fn, arity, args.length)
        next callable.call(*args, &block) if kwargs.empty?

        Scope.check_keywords(fn, accepted, kwargs)
        callable.call(*args, **kwargs, &block)
      end
    end

    # Bind the code modules `GG::Lib` collected at `lib.<key>`, or remove `lib` when there are none.
    #
    # @return [Boolean] whether a `lib` name was installed
    def self.install_lib
      Object.send(:remove_method, :lib) if Object.method_defined?(:lib)
      modules = Lib.registry
      return false if modules.empty?

      namespaces = Lib::Namespaces.new(modules)
      Object.send(:define_method, :lib) { namespaces }
      true
    end

    lift
  end
end
